### PR TITLE
Optimise snapshot deletion to speed up snapshot deletion and creation

### PR DESCRIPTION
**Pull request by ashking94**
### Description
Snapshot creation is distributed in nature. The snapshot creation operation is performed by the Data node holding the primary shard. Hence the total snapshot creation work is shared amongst all the data nodes in the cluster. On the contrary, the snapshot deletion is handled solely by active cluster manager. This can lead to excessively slow snapshots deletion when there are relative higher number of primary shards in the cluster.

In this PR, we have tried fixing this problem by creating a separate thread that is responsible for performing snapshot deletion or old shard gen cleanup during snapshot creation. The thread count has been set as 4x the number of allocated processor. The thread count is bounded between 64 and 256 to ensure that we have sufficient threads to get the deletion done and not too many threads that they start eating up from the connections of other remote store operations that may happen on the same cluster. 

### Check List
- [x] Functionality includes testing.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
Tests are pending as of now.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 1]
```diff
@@ -234,6 +236,7 @@ public ThreadPool(
         final int halfProcMaxAt5 = halfAllocatedProcessorsMaxFive(allocatedProcessors);
         final int halfProcMaxAt10 = halfAllocatedProcessorsMaxTen(allocatedProcessors);
         final int genericThreadPoolMax = boundedBy(4 * allocatedProcessors, 128, 512);
+        final int snapshotDeletionPoolMax = boundedBy(4 * allocatedProcessors, 64, 256);
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:239]: should it be `halfProc` else it can overwhelm the cluster manager ?

**Comment by ashking94** [comment]
> ❌ Gradle check result for [345a277](https://github.com/opensearch-project/OpenSearch/commit/345a277d578cceabb8ceb21f2fe4d8cb8f750666): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/46347/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky tests - https://github.com/opensearch-project/OpenSearch/issues/15600

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (ashking94) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:239]: These are only going to perform IO operations. I have also tested out on lower instance types and it works pretty well. On higher instance types, the impact is negligible.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 2]
```diff
@@ -251,6 +254,10 @@ public ThreadPool(
         builders.put(Names.REFRESH, new ScalingExecutorBuilder(Names.REFRESH, 1, halfProcMaxAt10, TimeValue.timeValueMinutes(5)));
         builders.put(Names.WARMER, new ScalingExecutorBuilder(Names.WARMER, 1, halfProcMaxAt5, TimeValue.timeValueMinutes(5)));
         builders.put(Names.SNAPSHOT, new ScalingExecutorBuilder(Names.SNAPSHOT, 1, halfProcMaxAt5, TimeValue.timeValueMinutes(5)));
+        builders.put(
+            Names.SNAPSHOT_DELETION,
+            new ScalingExecutorBuilder(Names.SNAPSHOT_DELETION, 1, snapshotDeletionPoolMax, TimeValue.timeValueMinutes(5))
+        );
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:260]: Lets include a changelog as users should know that deletion will now happen on a separate threadpool .

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:239]: have we tested with more number of shards as well ?

**Review by gbbafna — APPROVED** [review]

**Review by gbbafna — APPROVED** [review]

**Review by sachinpkale — APPROVED** [review]

**Pull request merged**